src/java/org/apache/cassandra/index/sai/view/IndexViewManager.java

```diff
@@ -0,0 +1,163 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.view;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.SSTableContext;
+import org.apache.cassandra.index.sai.SSTableIndex;
+import org.apache.cassandra.index.sai.StorageAttachedIndexGroup;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.utils.Pair;
+
+/**
+ * Maintain an atomic view for read requests, so that requests can read all data during concurrent compactions.
+ *
+ * All per-column {@link SSTableIndex} updates should be proxied by {@link StorageAttachedIndexGroup} to make
+ * sure per-sstable {@link SSTableContext} are in-sync.
+ */
+public class IndexViewManager
+{
+    private static final Logger logger = LoggerFactory.getLogger(IndexViewManager.class);
+    
+    private final IndexContext context;
+    private final AtomicReference<View> view = new AtomicReference<>();
+
+    public IndexViewManager(IndexContext context)
+    {
+        this(context, Collections.emptySet());
+    }
+
+    @VisibleForTesting
+    IndexViewManager(IndexContext context, Collection<SSTableIndex> indices)
+    {
+        this.context = context;
+        this.view.set(new View(context, indices));
+    }
+
+    public View getView()
+    {
+        return view.get();
+    }
+
+    /**
+     * Replaces old SSTables with new by creating new immutable view.
+     *
+     * @param oldSSTables A set of SSTables to remove.
+     * @param newSSTableContexts A set of SSTableContexts to add to tracker.
+     * @param validate if true, per-column index files' header and footer will be validated.
+     *
+     * @return A set of SSTables which have attached to them invalid index components.
+     */
+    public Set<SSTableContext> update(Collection<SSTableReader> oldSSTables, Collection<SSTableContext> newSSTableContexts, boolean validate)
```

Comment:
If you look at the whole chain of callers, `oldSSTables` could really be a `Set` rather than a collection. The only places where it's a `List`, it's an empty list. Then below, it seems like there's no point in creating `toRemove` if we already have a set and we're not modifying it?